### PR TITLE
Reviews

### PR DESCRIPTION
**Pull request by perioh**
This telegram bot allows you to add your Ukrzaliznytsia train ticket either with .pdf file or with link (the default way to share the ticket via mobile app).
The app will then remind you about your train 60, 30, 15 minutes prior to departure. If the train is being late - you will be notified about it, with approximate delay and approximate time of train arrival.
You can add multiple trains and be notified about each of them.

Program consists of 4 workspaces: 
- pdf_parser (parsing either bytes of .pdf train or filepath as TicketData struct with departure time and train number in it) 
- ukrzaliznytsia_parser (reqwest web client with ability to fetch page with delayed trains, parse its document object model and extract trains directions, numbers and delay time)
- telegram (for working with telegram bot via teloxide crate. Is used for fetching tickets, sent from users, and sending notifications about delays)
- database (database trait. I used HashSet as database, but this trait can be implemented for any other structure)

I decided to use workspaces, because I like the idea of each of them being kind of a "building blocks", allowing to combine them in desired way. This workspaces are separated and have no relation between each other, though they do share some common external crates.
Telegram bot link: [https://t.me/uz_reminder_bot](https://t.me/uz_reminder_bot) (is running on a server)

### CODE REVIEW

**Comment by BigglesworthCat** [comment]
A good, interesting and, most importantly, useful project. 
It is great that it involves different topics such as deserialisation of pdf files and development of a telegram bot. 
I like the organisation of the project modules and the way the code is written. It is really easy to read.

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+            telegram_worker(db, tg).await;
+        }
+    });
+
+    let mut notify_before_train_desc = NOTIFY_BEFORE_TRAIN;
+    notify_before_train_desc.sort_by(|a, b| b.cmp(a));
+
+    let mut notifications: HashMap<ChatId, HashSet<Duration>> = HashMap::new();
+
+    loop {
```
Comment (m1el) [inline src/main.rs:53]: this polling loop should be a separate function

Comment (imlvts) [inline src/main.rs:54]: this loop should be a separate function

**Comment by imlvts** [comment]
Overall, this looks good to me.
HTML parser can be done using CSS/XPath selectors, but doing it manually works as well.
Telegram state machine for user dialogue could have been used, but since the sate nearly non-existent, this is a decent solution.
The polling loop should be a separate function not to clutter the main function.

**Review by imlvts — APPROVED** [review]

**Comment by R0ckyRacc00n** [comment]
I like your idea. Looks as very useful tool that I can totally see myself using it. I don't see any issues other than those already written about 🦀

**Review by R0ckyRacc00n — APPROVED** [review]

**Review by BigglesworthCat — APPROVED** [review]

**Review by trepachkoDm — APPROVED** [review]
Overall a great program, back when Andi was approving this thread, I really liked the idea itself even then.

 I also want to mention that Andi has made a well structured code, I like his modularity system, it's well written and clear.

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    let kyiv_time = Kyiv.from_utc_datetime(&local_time.naive_utc());
+
+    kyiv_time
+}
+
+fn init_tracing() -> Result<WorkerGuard, Box<dyn Error>> {
+    let file_appender = RollingFileAppender::builder()
+        .rotation(Rotation::DAILY)
+        .filename_prefix("uzbot.log")
+        .build("logs")?;
```
Comment (TheBestTvarynka) [inline src/main.rs]: log file name should not be hardcoded

**Review by TheBestTvarynka — APPROVED** [review]
good code

File: src/main.rs (hunk 2 above)
Comment (perioh) [reply src/main.rs]: > log file name should not be hardcoded

done

**Pull request merged**